core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -138,8 +139,8 @@ class LogSegment(val log: FileMessageSet,
     val length = 
       maxOffset match {
         case None =>
-          // no max offset, just use the max size they gave unmolested
-          maxSize
+          // no max offset, just read until the max position
+          (maxPosition.getOrElse(size) - startPosition.position).toInt
```

Comment:
We should cap length by maxSize here, right? Also, I think it's probably simpler to not make maxPosition an option.